File: c++/clone-graph.cpp
//https://oj.leetcode.com/problems/clone-graph/

/**
 * Definition for undirected graph.
 * struct UndirectedGraphNode {
 *     int label;
 *     vector<UndirectedGraphNode *> neighbors;
 *     UndirectedGraphNode(int x) : label(x) {};
 * };
 */
class Solution {
public:
    UndirectedGraphNode *cloneGraph(UndirectedGraphNode *node) {
        if (node == NULL) return NULL;
        
        unordered_map<UndirectedGraphNode*, UndirectedGraphNode*> mapper;
        UndirectedGraphNode *newNode = new UndirectedGraphNode(node->label);
        mapper[node] = newNode;
        queue<UndirectedGraphNode*> q;
        q.push(node);
        
        while (!q.empty()) {
            UndirectedGraphNode *cur = q.front();
            q.pop();
            for (int i = 0; i < cur->neighbors.size(); i++) {
                UndirectedGraphNode *temp = cur->neighbors[i];
                if (mapper.find(temp) == mapper.end()) {
                    mapper[temp] = new UndirectedGraphNode(temp->label);
                    q.push(temp);
                }
                mapper[cur]->neighbors.push_back(mapper[temp]);
            }
        }

        return newNode;
    }
};
                

File: python/clone-graph.py
# Definition for a undirected graph node
# class UndirectedGraphNode:
#     def __init__(self, x):
#         self.label = x
#         self.neighbors = []

class Solution:
    # @param node, a undirected graph node
    # @return a undirected graph node
    def cloneGraph(self, node):
        if not node:
            return None

        queue = collections.deque([node])
        mapper = {id(node) : UndirectedGraphNode(node.label)}

        while queue:
            nd = queue.popleft()
            copy = mapper[id(nd)]
            for nei in nd.neighbors:
                if id(nei) not in mapper:
                    mapper[id(nei)] = UndirectedGraphNode(nei.label)
                    queue.append(nei)
                copy.neighbors.append(mapper[id(nei)])

        return mapper[id(node)]
